Add Home Assistant Vehicle

thecem opened the pull request:
This PR extends the Home Assistant vehicle adapter for EVCC to support all major and optional vehicle features, achieving feature parity with other adapters (e.g., VW, BMW). Key improvements include:

Support for start/stop charging via Home Assistant scripts/services.
Additional sensor support: target SOC, odometer, climatisation, max current, finish time, wakeup, etc.
Robust error handling for optional and missing sensors/services.
Refactored YAML template and configuration for flat parameter structure.
Comprehensive, up-to-date documentation and YAML example.
Compile-time interface checks for all supported features.
All changes are backward compatible. Unused fields are optional and can be omitted from the configuration. The adapter now enables seamless integration of Home Assistant vehicle data and control with EVCC.

thecem committed:
Create homeassistant.go

thecem committed:
Create homeassistant.yaml

thecem committed:
feat(vehicle): add homeassistant

thecem committed:
feat(vehicle): add homeassistant

thecem committed:
feat(vehicle): add homeassistant

Correct description

thecem committed:
fix repeated logic and normalizing the state string in Climater before comparison.

thecem commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,347 @@
+}
+
+// Climater liefert true, wenn Klimatisierung aktiv ist (optional)
+func (v *HomeAssistant) Climater() (bool, error) {
```
fixed

thecem committed:
Fix test and UI error

Remove the keys cache, password, user, and vin from your homeassistant.yaml template (and any test/demo configs using this template).
Only include keys that are actually supported by the haConfig struct and the Home Assistant vehicle implementation.

thecem committed:
fix test

Remove both - preset: vehicle-base from the params section and {{- include "vehicle-base" . }} from the render block.

thecem committed:
fix test

remove preset vehicle-base

thecem committed:
fis test

remove 2 blank lines

thecem committed:
fix test edit token and add dummy-token

andig commented on templates/definition/vehicle/homeassistant.yaml:
```diff
@@ -0,0 +1,139 @@
+  token: {{ or .token "dummy-token" }}
+  sensors:
+    soc: {{ .sensors_soc }}
+    {{- if .sensors_range }}
```
All if statements can simply be omitted

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+
+// Calls script.<name> as a service call without payload
+func (v *HomeAssistant) callScript(script string) error {
+	if script == "" {
```
Use decorators instead

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+
+func (v *HomeAssistant) getIntSensor(entity string) (int64, error) {
+	if entity == "" {
+		return 0, api.ErrNotAvailable
```
Dito

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+
+func (v *HomeAssistant) getTimeSensor(entity string) (time.Time, error) {
+	if entity == "" {
+		return time.Time{}, api.ErrNotAvailable
```
Etc…

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+
+// Soc returns state of charge [%]
+func (v *HomeAssistant) Soc() (float64, error) {
+	return v.getFloatSensor(v.conf.Sensors.Soc)
```
Dito..  all thes eoptional readings MUST be decorated

thecem replied:
decorated

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+		embed:        &cc.embed,
+		Helper:       request.NewHelper(log),
+		conf:         cc,
+		startScript:  cc.Services.Start,
```
Why are both conf and conf.services copied?

thecem replied:
solved

andig replied:
Du musst die nicht alle kommentieren ;). Wenn sich das diff ändert macht GH die automatisch zu.

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+
+// Calls /api/states/<entity> and returns .state
+func (v *HomeAssistant) getState(entity string) (string, error) {
+	uri := fmt.Sprintf("%s/api/states/%s", strings.TrimSuffix(v.conf.Host, "/"), entity)
```
Trim host once on startzp, not per request

thecem replied:
getrimmt

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+	if ts, err := strconv.ParseInt(s, 10, 64); err == nil {
+		return time.Unix(ts, 0), nil
+	}
+	t, err := time.Parse(time.RFC3339, s)
```
return ..

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+}
+
+// StartCharge triggers the start charging script
+func (v *HomeAssistant) StartCharge() error { return v.callScript(v.startScript) }
```
Make method private

thecem replied:
done

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+// StartCharge triggers the start charging script
+func (v *HomeAssistant) StartCharge() error { return v.callScript(v.startScript) }
+
+// StopCharge triggers the stop charging script
```
Dito

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+// WakeUp triggers the wakeup script (optional)
+func (v *HomeAssistant) WakeUp() error {
+	if v.wakeupScript == "" {
+		return api.ErrNotAvailable
```
Decorate

thecem replied:
done

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,320 @@
+// Compile-time checks
+// -----------------------------
+
+var (
```
To be removed

thecem replied:
removed

andig commented on templates/definition/vehicle/homeassistant.yaml:
```diff
@@ -0,0 +1,139 @@
+template: homeassistant
```
group: generic

thecem committed:
fix all comments

thecem committed:
fix all comments and blank lines ;-)

thecem committed:
try to fix all comments

thecem committed:
remove var 

// -----------------------------
// Compile-time checks
// -----------------------------
var (
	_ api.Vehicle            = (*HomeAssistant)(nil)
	_ api.VehicleRange       = (*HomeAssistant)(nil)
	_ api.ChargeState        = (*HomeAssistant)(nil)
	_ api.VehicleOdometer    = (*HomeAssistant)(nil)
	_ api.SocLimiter         = (*HomeAssistant)(nil)
	_ api.Resurrector        = (*HomeAssistant)(nil)
	_ api.VehicleClimater    = (*HomeAssistant)(nil)
	_ api.VehicleFinishTimer = (*HomeAssistant)(nil)
)

thecem commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,352 @@
+}
+
+// Status returns evcc charge status (optional)
+func (v *HomeAssistant) Status() (api.ChargeStatus, error) {
```
fixed

thecem commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,352 @@
+	return resp.State, nil
+}
+
+// Calls script.<name> as a service call without payload
+func (v *HomeAssistant) callScript(script string) error {
```
fixed

thecem committed:
fix comments of Sourcery-AI

https://github.com/evcc-io/evcc/pull/22594#pullrequestreview-3069904911

andig commented on templates/definition/vehicle/homeassistant.yaml:
```diff
@@ -0,0 +1,119 @@
+template: homeassistant
+products:
+  - description:
+      generic: Home Assistant Vehicle Integration
+      de: Home Assistant Fahrzeug Integration
+      en: Home Assistant Vehicle Integration
```
```suggestion
      de: Home Assistant Fahrzeug
      en: Home Assistant Vehicle
```

andig committed:
Update templates/definition/vehicle/homeassistant.yaml

andig committed:
Simplify

andig committed:
Simplify

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,360 @@
+	return v.callScript(v.conf.Services.Stop)
+}
+
+// ChargeEnable enables or disables charging via the UI toggle.
```
is UI relevant here?

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,360 @@
+}
+
+// WakeUp triggers the Home Assistant wakeup script (if configured).
+// Returns api.ErrNotAvailable if no wakeup script is set.
```
```suggestion
```

andig commented on templates/definition/vehicle/homeassistant.yaml:
```diff
@@ -0,0 +1,118 @@
+  - name: soc
+    description:
+      de: Ladezustand [%] (required)
+      en: State of charge [%] (required)
```
remove all required/option descriptions. this is managed by the required key.

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,360 @@
+		vehicleFinish = func() (time.Time, error) { return base.getTimeSensor(cc.Sensors.FinishTime) }
+	}
+	if cc.Sensors.MaxCurrent != "" {
+		// MaxCurrent is not a setter in HomeAssistant, so leave nil or implement if needed
```
What does this mean? How is the method decorated?

andig replied:
Depends. What's the intention here?

thecem replied:
Used 1 reference
The intention depends on how your Home Assistant integration is set up and what EVCC expects:

If you want to read the current max charging current (getter), you should implement and provide a function for GetMaxCurrent (which you already have).
If you want to set the max charging current (setter), you should implement and provide a function for SetMaxCurrent (which you already have).
The MaxCurrent field in the EVCC vehicle interface is typically a getter (reads the current limit). In your Home Assistant setup, this is usually covered by GetMaxCurrent (reading from a sensor/entity that reflects the current limit set in Home Assistant).
In summary:

If your Home Assistant exposes a sensor for the current max current (e.g., sensor.id4_actual_max_current), use GetMaxCurrent.
If you want to allow EVCC to set the max current, implement SetMaxCurrent.
If you do not have a sensor for reading the max current, leave the getter nil.
If you do not want to allow setting, leave the setter nil.
Typical usage in EVCC:

Provide a getter if you want to display or use the current max current in automations.
Provide a setter if you want EVCC to be able to change the max current via Home Assistant.
Let me know your Home Assistant setup and what you want to achieve, and I can generate the correct code for your use case!  

Do we need this? for evcc in some vehicle cenarios?

thecem replied:
removed

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,360 @@
+	return resp.State, nil
+}
+
+/*
+callScript calls script.<name> as a Home Assistant service call without payload.
+
+Note: This function currently does not support passing parameters to scripts.
+If you need to call scripts with parameters, payload support should be added in the future.
+
+TODO: Add support for script parameters/payloads if needed.
+*/
```
```suggestion
// callScript calls script.<name> as a Home Assistant service call without payload
```

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,360 @@
+
+// MaxCurrent returns the maximum charging current (optional)
+func (v *HomeAssistant) MaxCurrent() (int64, error) {
+	return v.getIntSensor(v.conf.Sensors.MaxCurrent)
```
is this a setter or getter?

andig replied:
there are 3 current methods now- 1 seems unnecessary

thecem commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,360 @@
+
+type haConfig struct {
+	embed    `mapstructure:",squash"`
+	Host     string // http://ha:8123
```
uri

andig replied:
No comments required, just change it as necessary ;)

thecem committed:
fixed comment

removed maxcurrent and getmaxcurrent
host -> uri
remove log
remove descriptions
change methodenames

thecem committed:
fix option/requiered rem (get)maxcurrent

thecem committed:
Merge branch 'master' into homeassistant-vehicle

thecem committed:
add custom HTTP transport (tokenTransport)

A custom HTTP transport (tokenTransport) was added to always inject the Home Assistant token as an Authorization header for every request. This ensures all API calls are authenticated automatically, making the code cleaner and more robust. Manual header injection in each request is no longer needed.

thecem committed:
removed unused funktions

thecem committed:
Merge branch 'master' into homeassistant-vehicle

andig committed:
Apply suggestion from @andig

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,326 @@
+	// decorate all features
+	return decorateVehicle(
+		base,
+		base.getLimitSoc,
+		base.status,
```
Die gehören analog oben in Variablen

thecem committed:
tryed to fix all comments

thecem committed:
Merge branch 'master' into homeassistant-vehicle

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,335 @@
+	"github.com/evcc-io/evcc/util/transport"
+)
+
+// Public wrappers to satisfy api.Vehicle interface, call function variables if set
+func (v *HomeAssistant) Soc() (float64, error) {
+	if v.socFunc != nil {
+		return v.socFunc()
+	}
+	return 0, api.ErrNotAvailable
+}
+func (v *HomeAssistant) Range() (int64, error) {
+	if v.rangeFunc != nil {
+		return v.rangeFunc()
+	}
+	return 0, api.ErrNotAvailable
+}
+func (v *HomeAssistant) Odometer() (float64, error) {
+	if v.odoFunc != nil {
+		return v.odoFunc()
+	}
+	return 0, api.ErrNotAvailable
+}
+func (v *HomeAssistant) Climater() (bool, error) {
+	if v.climaterFunc != nil {
+		return v.climaterFunc()
+	}
+	return false, api.ErrNotAvailable
+}
+func (v *HomeAssistant) FinishTime() (time.Time, error) {
+	if v.finishFunc != nil {
+		return v.finishFunc()
+	}
+	return time.Time{}, api.ErrNotAvailable
+}
+func (v *HomeAssistant) LimitSoc() (float64, error) {
+	if v.limitSocFunc != nil {
+		return v.limitSocFunc()
+	}
```
Bitte alles raus. KEIN public Funktionen für optionales Verhalten.

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,335 @@
+	conf haConfig
+	uri  string
+
+	socFunc      func() (float64, error)
+	rangeFunc    func() (int64, error)
+	odoFunc      func() (float64, error)
+	climaterFunc func() (bool, error)
+	finishFunc   func() (time.Time, error)
+	limitSocFunc func() (float64, error)
```
Raus

andig commented on vehicle/homeassistant.go:
```diff
@@ -0,0 +1,335 @@
+	return v.getTimeSensor(v.conf.Sensors.FinishTime)
+}
+
+// GetLimitSoc implements api.SocLimiter (Decorator-Interface)
+func (v *HomeAssistant) GetLimitSoc() (int64, error) {
+	if v.limitSocFunc != nil {
+		val, err := v.limitSocFunc()
+		return int64(val), err
+	}
+	return 0, api.ErrNotAvailable
+}
```
Raus

thecem committed:
tryed to fix comments

Die Home Assistant Vehicle-Implementierung erfüllt jetzt alle Anforderungen von @andig:

🎯 True Decorator Pattern Implementation
✅  Directive für automatische Decorator-Code-Generierung
✅ Conditional Feature Exposure - nur konfigurierte Features werden als Interface-Methoden verfügbar gemacht
✅ Private Helper Methods statt public interface methods für optionale Features
✅ Clean Interface Separation zwischen mandatory (Soc()) und optional features
🎯 Backward Compatibility
✅ Services Support - unterstützt sowohl alte services: Struktur als auch neue flache Struktur
✅ Legacy Configuration wird weiterhin unterstützt
✅ Graceful Fallbacks für nicht konfigurierte Features
🎯 Optimized Interface Compliance
✅ api.Vehicle (mandatory) - Soc() immer verfügbar
✅ api.SocLimiter - GetLimitSoc() nur wenn limitSoc konfiguriert
✅ api.VehicleRange - Range() nur wenn range konfiguriert
✅ api.VehicleOdometer - Odometer() nur wenn odometer konfiguriert
✅ api.VehicleClimater - Climater() nur wenn climater konfiguriert
✅ api.ChargeState - Status() nur wenn status konfiguriert
✅ api.ChargeController - ChargeEnable() nur wenn services oder enable konfiguriert
✅ api.Resurrector - WakeUp() nur wenn wakeup konfiguriert
✅ api.VehicleFinishTimer - FinishTime() nur wenn finishTime konfiguriert

thecem committed:
flaten struct

thecem committed:
correct wakeup

andig commented:
Könntest du den bitte auf den Stand von gestern zurück drehen? Alles danach geht in die falsche Richtung.

thecem committed:
Merge branch 'master' into homeassistant-vehicle

thecem committed:
fix

thecem committed:
Update homeassistant.go

thecem committed:
Update homeassistant.go

thecem committed:
Merge branch 'master' into homeassistant-vehicle

thecem committed:
Update homeassistant.yaml

thecem committed:
Update homeassistant.go

thecem committed:
Update homeassistant.yaml

andig committed:
FIX

andig committed:
wip

andig committed:
wip

andig committed:
Simplify

andig committed:
Update vehicle/homeassistant.go

andig committed:
wip

andig committed:
wip

andig committed:
wip

andig committed:
wip

thecem committed:
service call uri adapted

thecem committed:
adapted actual struct

sensors:
  soc:

thecem commented:
Die uri und die .yaml mussten noch angepasst werden. -> GEHT!

<details>
  <summary>evcc vehicle ha_vehicle output</summary>

```bash
vscode ➜ /workspaces/evcc (master) $ ./evcc vehicle ha_vehicle
[main  ] INFO 2025/08/01 18:36:26 using config file: /workspaces/evcc/evcc.yaml
[db    ] INFO 2025/08/01 18:36:26 using sqlite database: /home/vscode/.evcc/evcc.db
Soc:            68%
Capacity:       77.0kWh
Charge status:  A
Range:          231km
Odometer:       34794km
Climate active: false
Limit Soc:      80%
```

</details>

./evcc vehicle ha_vehicle --start/stop/wakeup -> OK

thecem commented:
https://github.com/evcc-io/evcc/discussions/22676
